File: src/cuda/gpu_distance.rs
use std::ffi::c_void;
use std::ptr;

use crate::bindings::FaissMetricType;
use crate::cuda::gpu_resources::GpuResourcesProvider;
use crate::error::{check_return_code, Result};

pub struct GpuDistanceParams {
    inner: *mut c_void,
}

impl Drop for GpuDistanceParams {
    fn drop(&mut self) {
        if !self.inner.is_null() {
            unsafe { crate::extension::cuda::gpu_distance_params_free(self.inner) }
        }
    }
}

impl GpuDistanceParams {
    pub fn new(metric: FaissMetricType, dims: i32, device: i32) -> Result<Self> {
        let mut inner = ptr::null_mut();
        check_return_code(unsafe {
            crate::extension::cuda::gpu_distance_params_new(&mut inner, metric as i32, dims, device)
        })?;
        Ok(Self { inner })
    }

    pub fn set_vectors(&mut self, vectors: &[f32]) {
        unsafe {
            crate::extension::cuda::gpu_distance_params_set_vectors(self.inner, vectors.as_ptr())
        }
    }

    pub fn set_queries(&mut self, queries: &[f32]) {
        unsafe {
            crate::extension::cuda::gpu_distance_params_set_queries(self.inner, queries.as_ptr())
        }
    }

    pub fn set_k(&mut self, k: i32) {
        unsafe { crate::extension::cuda::gpu_distance_params_set_k(self.inner, k) }
    }

    pub fn set_num_vectors(&mut self, num_vectors: i32) {
        unsafe {
            crate::extension::cuda::gpu_distance_params_set_num_vectors(self.inner, num_vectors)
        }
    }

    pub fn set_num_queries(&mut self, num_queries: i32) {
        unsafe {
            crate::extension::cuda::gpu_distance_params_set_num_queries(self.inner, num_queries)
        }
    }

    pub fn set_results(&mut self, results: &mut [f32]) {
        unsafe {
            crate::extension::cuda::gpu_distance_params_set_results(
                self.inner,
                results.as_mut_ptr(),
            )
        }
    }

    pub fn set_indices(&mut self, indices: &mut [i64]) {
        unsafe {
            crate::extension::cuda::gpu_distance_params_set_indices(
                self.inner,
                indices.as_mut_ptr(),
            )
        }
    }

    pub fn compute(&self, resources: &impl GpuResourcesProvider) -> Result<()> {
        check_return_code(unsafe {
            crate::extension::cuda::gpu_distance_params_compute(
                self.inner,
                resources.inner() as *mut _,
            )
        })
    }

    pub fn get_dims(&self) -> i32 {
        let mut dims = 0;
        unsafe { crate::extension::cuda::gpu_distance_params_get_dims(self.inner, &mut dims) }
        dims
    }

    pub fn get_k(&self) -> i32 {
        let mut k = 0;
        unsafe { crate::extension::cuda::gpu_distance_params_get_k(self.inner, &mut k) }
        k
    }

    pub fn get_num_vectors(&self) -> i32 {
        let mut num = 0;
        unsafe { crate::extension::cuda::gpu_distance_params_get_num_vectors(self.inner, &mut num) }
        num
    }

    pub fn get_num_queries(&self) -> i32 {
        let mut num = 0;
        unsafe { crate::extension::cuda::gpu_distance_params_get_num_queries(self.inner, &mut num) }
        num
    }
}

File: src/extension/gpu_distance.cpp
#ifdef USE_CUDA

#include <faiss/gpu/GpuDistance.h>
#include <faiss/gpu/StandardGpuResources.h>
#include "gpu_distance.hpp"

struct GpuDistanceParamsInternal {
    faiss::gpu::GpuDistanceParams params;
    int device;
};

int gpu_distance_params_new(GpuDistanceParams **params, int metric, int dims, int device)
{
    try
    {
        auto *internal = new GpuDistanceParamsInternal();
        internal->device = device;
        internal->params.metric = static_cast<faiss::MetricType>(metric);
        internal->params.dims = dims;
        *params = (GpuDistanceParams *)internal;
        return 0;
    }
    catch (std::exception &e)
    {
        return -1;
    }
}

void gpu_distance_params_free(GpuDistanceParams *params)
{
    delete (GpuDistanceParamsInternal *)params;
}

void gpu_distance_params_get_dims(GpuDistanceParams *params, int *dims)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    *dims = internal->params.dims;
}

void gpu_distance_params_set_vectors(GpuDistanceParams *params, const float *vectors)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    internal->params.vectors = vectors;
}

void gpu_distance_params_get_k(GpuDistanceParams *params, int *k)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    *k = internal->params.k;
}

void gpu_distance_params_set_k(GpuDistanceParams *params, int k)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    internal->params.k = k;
}

void gpu_distance_params_get_num_vectors(GpuDistanceParams *params, int *num_vectors)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    *num_vectors = internal->params.numVectors;
}

void gpu_distance_params_get_num_queries(GpuDistanceParams *params, int *num_queries)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    *num_queries = internal->params.numQueries;
}

void gpu_distance_params_set_num_vectors(GpuDistanceParams *params, int num_vectors)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    internal->params.numVectors = num_vectors;
}

void gpu_distance_params_set_queries(GpuDistanceParams *params, const float *queries)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    internal->params.queries = queries;
}

void gpu_distance_params_set_num_queries(GpuDistanceParams *params, int num_queries)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    internal->params.numQueries = num_queries;
}

void gpu_distance_params_get_results(GpuDistanceParams *params, float **results)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    *results = internal->params.outDistances;
}

void gpu_distance_params_set_results(GpuDistanceParams *params, float *results)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    internal->params.outDistances = results;
}

void gpu_distance_params_get_indices(GpuDistanceParams *params, int64_t **indices)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    *indices = internal->params.outIndices;
}

void gpu_distance_params_set_indices(GpuDistanceParams *params, int64_t *indices)
{
    auto *internal = (GpuDistanceParamsInternal *)params;
    internal->params.outIndices = indices;
}

int gpu_distance_params_compute(GpuDistanceParams *params, GpuResources *resources)
{
    try
    {
        auto *internal = (GpuDistanceParamsInternal *)params;
        auto *res = (faiss::gpu::GpuResources *)resources;
        faiss::gpu::bfKnn(*res, internal->device, internal->params);
        return 0;
    }
    catch (std::exception &e)
    {
        return -1;
    }
}

#endif